### DISCUSSION TITLE
Large amount of deletions/update requests on large amount of data.

### DISCUSSION THREAD

**Discussion by gulnara** [ClickHouse/ClickHouse#56021] [Q&A]
Hi, as part of support for GDPR and BCR we need to be able to delete/redact data in ClickHouse. Currently we are running into an issue with too many mutations where ClickHouse slows down considerably when receiving deletion and redaction requests.
We use lightweight deletes for deletions and update queries for redactions.
What is a ClickHouse recommended approach to deleting/updating large amounts of data under a large amount of load.

**Comment by den-crane** [comment]
The recommendation is to run small number of delete operations. 
You need to accumulate ids of rows which should be deleted and delete them using single operation once a week/day.

For example instead of

```
delete from where id=?;
delete from where id=?;
....
delete from where id=?;
```

you can do

```
insert into some_replicated_table values (id);
insert into some_replicated_table values (id);
...
insert into some_replicated_table values (id);

delete from where id in (select id from some_replicated_table);
```

also see:
https://kb.altinity.com/altinity-kb-queries-and-syntax/row_policy_using_dictionary/

https://kb.altinity.com/altinity-kb-queries-and-syntax/delete-via-tombstone-column/

**Reply by gulnara** [reply]
@den-crane , thank you for the recommendation. Two follow up questions:
1) how would it limit number of mutations?
2) would you recommend same approach for updates?

**Reply by den-crane** [reply]
Instead of many mutations, only one mutation will be executed? 
For simplicity you may think that delete mutation rewrites a table without deleted rows, lightweight delete and update rewrite the whole column (all rows). So basically one fat mutation which deletes half of rows in the table does the same work as one mutation which deletes one row. So 1000 small mutations is 1000 times more heavy than one big mutation.

**Reply by den-crane** [reply]
```
create table test (a Int64, s String) Engine=MergeTree order by a as select *, * from numbers(1e7);

select count() from test;
┌──count()─┐
│ 10000000 │
└──────────┘

delete from test where a = 42; 
0 rows in set. Elapsed: 0.157 sec.  -- one row delete

select count() from test;
┌─count()─┐
│ 9999999 │
└─────────┘


delete from test where a % 2 = 0;
0 rows in set. Elapsed: 0.214 sec.  -- 5 000 000 rows delete

select count() from test;
┌─count()─┐
│ 5000000 │
└─────────┘
```

as you can see 5 000 000 rows delete almost equal one row delete

**Reply by gulnara** [reply]
This makes a lot of sense! Thank you for example.

**Reply by zillionare** [reply]
How could we use same strategy to do update? I've tried several ways but failed.

Given the following table:

```sql
CREATE TABLE if not exists bars_day
(
    `frame` Date,
    `symbol` LowCardinality(String),
    `open` Float32 DEFAULT -1 CODEC(Delta, ZSTD),
    `high` Float32 DEFAULT -1 CODEC(Delta, ZSTD),
    `low` Float32 DEFAULT -1 CODEC(Delta, ZSTD),
    `close` Float32 DEFAULT -1 CODEC(Delta, ZSTD),
    `factor` Float64 DEFAULT -1 CODEC(Delta, ZSTD),
)
ENGINE = MergeTree()
ORDER BY (frame, symbol)
```

frame and symbol determines unique row here.
as of time we inserted data to bars_day, factor may not be available, so we have to update that part later.  If I managed to write factor into another table, say, temp_for_factor_merge:

```sql
create table temp_for_merge_factor (
    `frame` Date,
    `symbol` LowCardinality(String),
    `factor` Float64
)
ENGINE = MergeTree
ORDER BY (frame, symbol)
```

I have tried the following sql but both failed:

```sql
alter table bars_day update bars_day.factor = temp_for_merge_factor.factor FROM
bars_day
inner join temp_for_merge_factor
on bars_day.symbol=temp_for_merge_factor.symbol
and bars_day.frame=temp_for_merge_factor.frame
```
or 

```sql
alter table tests.bars_day update factor = (
select frame, symbol, factor from tests.temp_for_merge_factor inner join tests.bars_day on tests.bars_day.symbol=tests.temp_for_merge_factor.symbol
)

```

Can anyone give an example on how to do build update like deletion?

**Reply by MSchutt** [reply]
It is not possible to do a subquery with the `ALTER TABLE` syntax. What you'd need to do is build up either a dictionary (https://clickhouse.com/docs/en/sql-reference/dictionaries) or a join table (https://clickhouse.com/docs/en/engines/table-engines/special/join) and then you can use `dictGet` / `joinGet` functions like such:

```sql
CREATE TABLE tests.bars_update (
   frame Date,
   symbol LowCardinality(String),
   factor Float64
)
— Setup the in-memory join table with the keys: symbol, frame
Engine = JOIN(ANY, LEFT, symbol, frame);
```
Next, insert the update data into the join table.

After that, you can issue the command to update your main table using the join table:
```sql
ALTER TABLE tests.bars_day
— Uses symbol + frame to lookup in the join table and returns the 'factor' variable
UPDATE factor = joinGet('tests.bars_update', 'factor', symbol, frame)
WHERE isNotNull(joinGetOrNull('tests.bars_update', 'factor', symbol, frame));
```

It's important to note that if you're using a cluster, the join table is not automatically replicated across the cluster. Consequently, you would need to create the table and insert the data on each server individually.
